vm/src/stdlib/itertools.rs

```diff
@@ -736,6 +737,140 @@ impl PyItertoolsTee {
     }
 }
 
+#[pyclass]
+#[derive(Debug)]
+struct PyIterToolsProduct {
+    pools: Vec<Vec<PyObjectRef>>,
+    idxs: RefCell<Vec<usize>>,
+    cur: RefCell<usize>,
+    sizes: Vec<usize>,
+    stop: RefCell<bool>,
+}
+
+impl PyValue for PyIterToolsProduct {
+    fn class(vm: &VirtualMachine) -> PyClassRef {
+        vm.class("itertools", "product")
+    }
+}
+
+#[derive(FromArgs)]
+struct ProductArgs {
+    #[pyarg(keyword_only, optional = true)]
+    repeat: OptionalArg<PyIntRef>,
+}
+
+#[pyimpl]
+impl PyIterToolsProduct {
+    #[pymethod(name="__new__")]
+    #[allow(clippy::new_ret_no_self)]
+    fn new(
+        _cls: PyClassRef,
+        iterables: Args<PyObjectRef>,
+        args: ProductArgs,
+        vm: &VirtualMachine,
```

Comment:
You can do `PyResult<PyRef<Self>>`, call `.into_ref_with_type(vm, cls)` and then you don't need the clippy `allow`